pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/proxywatcher.go

```diff
@@ -0,0 +1,111 @@
+package delegate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+
+	oauthapi "github.com/openshift/oauth-apiserver/pkg/oauth/apis/oauth"
+)
+
+// OAuthAccessTokenWatcher implements a Watcher interface to watch OAuthAccessTokens
+// and filter them to match only those for the current user
+type OAuthAccessTokenWatcher struct {
+	wrappedWatcher watch.Interface
+	outgoing       chan watch.Event
+	incoming       <-chan watch.Event
+	stopCh         chan struct{}
+
+	username string
+
+	mutex   sync.Mutex
+	stopped bool
+}
+
+var _ watch.Interface = &OAuthAccessTokenWatcher{}
+
+// NewOAuthAccessTokenWatcher creates new OAuthAccessTokenWatcher by wrapping a channel of watched
+// oauthaccesstokens. The username parameter allows filtering the oauthaccesstokens from the incoming
+// channel for that specific user.
+func newOAuthAccessTokenWatcher(wrappedWatcher watch.Interface, username string) *OAuthAccessTokenWatcher {
+	return &OAuthAccessTokenWatcher{
+		wrappedWatcher: wrappedWatcher,
+		incoming:       wrappedWatcher.ResultChan(),
+		outgoing:       make(chan watch.Event),
+		stopCh:         make(chan struct{}),
+		stopped:        false,
+
+		username: username,
+	}
+}
+
+// Stop implements Interface
+func (w *OAuthAccessTokenWatcher) Stop() {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	if !w.stopped {
+		w.stopped = true
+		w.wrappedWatcher.Stop()
+		close(w.stopCh)
+	}
+}
+
+// Stopping returns true if Stop() has been called
+func (w *OAuthAccessTokenWatcher) Stopping() bool {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	return w.stopped
+}
+
+// ResultChan returns channel for filtered results
+func (w *OAuthAccessTokenWatcher) ResultChan() <-chan watch.Event {
+	return w.outgoing
+}
+
+// StopChan returns stop channel
+func (w *OAuthAccessTokenWatcher) StopChan() <-chan struct{} {
```

Comment:
is this method used ?

Reply:
I copied the interface of the `apimachinery/pkg/watch.ProxyWatcher` struct, but it looks like this and `Stopping` may indeed be safe to remove